R/return_linter.R

```diff
@@ -0,0 +1,206 @@
+#' Return linter
+#'
+#' This linter checks for explicit [return()] at the end of a function
+#'
+#' @param use_implicit_returns Whether to use implicit or explicit returns
+#'
+#' @param additional_allowed_func Names of additional functions that are
+#'  accepted as return if `!use_implicit_returns`
+#'
+#' @param additional_side_effect_func Names of additional functions that are
+#'  not checked for an explicit retun if `!use_implicit_returns`
+#'
+#' @param use_runit Whether to ignore Runit like functions or not
```

Comment:
I would leave this argument out, the feature is already present by using the function whitelist arguments.

Reply:
In my opinion, we can also do without this parameter.  However, it cannot be replicated with the other two, because the way it was implemented, all functions that start with “Test” were ignored.  I wanted to keep that with this parameter, even though I'm not a supporter of this functionality.  I would also prefer without this parameter.

Reply:
Yea, it's hard to capture the exemption for `Test*` functions with an allowlist. It's hardly practical for users to supply every single `Test*` function in their suite, and I don't think there's much need for a pattern based argument accepting `Test*` (glob) or `^Test.*` (regex). I think `use_runit` or `allow_runit` is the best bet here.

Reply:
I'm not familiar with RUnit but would it be an option to directory-exclude the RUnit test files instead?

Reply:
not really, no. (1) there's not really a consistent directory structure or file name pattern for RUnit tests (2) it won't work in distributed setup where per-directory settings are not possible (even if eventually supported by lintr itself).

another option is to hack it such that if (any RUnit functions exempted) then "also exempt Test* functions

Reply:
We could also add a parameter "function_exclusion_prefixes" . Then one could fully emulate the functionality without adapting the linter too much to a package.

Reply:
What about adding rudimentary globbing support to `except=`? It could detect `*` at the end of exceptions and convert those entries to prefix-exclusions.

Reply:
Something along those lines could work. OTOH I worry about this PR + discussion getting too long.

WDYT about keeping a simple allowlist for this initial PR and handling extensions as a follow-up?

Reply:
#2335